crates/ruff_linter/src/rules/numpy/rules/numpy_2_0_deprecation.rs

```diff
@@ -701,3 +707,233 @@ pub(crate) fn numpy_2_0_deprecation(checker: &mut Checker, expr: &Expr) {
+        attribute_searcher.visit_stmt(stmt);
+        if attribute_searcher.found_attribute {
+            return true;
+        }
+    }
+    false
+}
+
+/// AST visitor that searches an AST tree for [`ast::ExprAttribute`] nodes
+/// that match a certain [`QualifiedName`].
+struct AttributeSearcher<'a> {
+    attribute_to_find: QualifiedName<'a>,
+    semantic: &'a SemanticModel<'a>,
+    found_attribute: bool,
+}
+
+impl<'a> AttributeSearcher<'a> {
+    fn new(attribute_to_find: QualifiedName<'a>, semantic: &'a SemanticModel<'a>) -> Self {
+        Self {
+            attribute_to_find,
+            semantic,
+            found_attribute: false,
+        }
+    }
+}
+
+impl Visitor<'_> for AttributeSearcher<'_> {
+    fn visit_expr(&mut self, expr: &'_ Expr) {
+        if self.found_attribute {
+            return;
+        }
+        if expr.is_attribute_expr()
+            && self
+                .semantic
+                .resolve_qualified_name(expr)
+                .is_some_and(|qualified_name| qualified_name == self.attribute_to_find)
+        {
+            self.found_attribute = true;
+            return;
+        }
+        ast::visitor::walk_expr(self, expr);
+    }
+}
+
+/// Given an [`ast::StmtTry`] node, does the `try` branch of that node
+/// contain any [`ast::StmtImportFrom`] nodes that indicate the numpy
+/// member is being imported from the non-deprecated location?
+fn try_block_contains_undeprecated_import(
+    try_node: &ast::StmtTry,
+    replacement_details: &Details,
+) -> bool {
+    let Details::AutoImport {
+        path,
+        name,
+        compatibility: _,
+    } = replacement_details
+    else {
+        return false;
+    };
+    let mut import_searcher = ImportSearcher::new(path, name);
+    for stmt in &try_node.body {
+        import_searcher.visit_stmt(stmt);
+        if import_searcher.found_import {
+            return true;
+        }
+    }
+    false
+}
+
+/// AST visitor that searches an AST tree for [`ast::StmtImportFrom`] nodes
+/// that match a certain [`QualifiedName`].
+struct ImportSearcher<'a> {
+    module: &'a str,
+    name: &'a str,
+    found_import: bool,
+}
+
+impl<'a> ImportSearcher<'a> {
+    fn new(module: &'a str, name: &'a str) -> Self {
+        Self {
+            module,
+            name,
+            found_import: false,
+        }
+    }
+}
+
+impl Visitor<'_> for ImportSearcher<'_> {
+    fn visit_stmt(&mut self, stmt: &ast::Stmt) {
+        if self.found_import {
+            return;
+        }
+        if let ast::Stmt::ImportFrom(ast::StmtImportFrom { module, names, .. }) = stmt {
+            if module.as_ref().is_some_and(|module| module == self.module)
+                && names
+                    .iter()
+                    .any(|ast::Alias { name, .. }| name == self.name)
+            {
+                self.found_import = true;
+                return;
+            }
+        }
+        ast::visitor::walk_stmt(self, stmt);
+    }
+}
```

Comment:
Nit: Add an empty `visit_expression` to avoid traversing **into** expressions or implement `StatementVisitor`